python/cuml/experimental/fil/fil.pyx

```diff
@@ -309,17 +324,75 @@ cdef class ForestInference_impl():
             # TODO(wphicks): Handle incorrect dtype/device/layout in C++
             preds.index = in_arr.index
         cdef raft_proto_device_t out_dev
-        if preds.is_device_accessible:
-            if (
-                GlobalSettings().device_type == DeviceType.host
-                and preds.is_host_accessible
-            ):
-                out_dev = raft_proto_device_t.cpu
-            else:
-                out_dev = raft_proto_device_t.gpu
+        out_dev = get_device_type(preds)
+        out_ptr = preds.ptr
+
+        cdef optional[uint32_t] chunk_specification
+        if chunk_size is None:
+            chunk_specification = nullopt
+        else:
+            chunk_specification = <uint32_t> chunk_size
+
+        if model_dtype == np.float32:
+            self.model.predict[float](
+                self.raft_proto_handle,
+                predict_t.predict,
+                <float*> out_ptr,
+                <float*> in_ptr,
+                n_rows,
+                out_dev,
+                in_dev,
+                chunk_specification
+            )
         else:
-            out_dev = raft_proto_device_t.cpu
+            self.model.predict[double](
+                self.raft_proto_handle,
+                predict_t.predict,
+                <double*> out_ptr,
+                <double*> in_ptr,
+                n_rows,
+                in_dev,
+                out_dev,
+                chunk_specification
+            )
+
+        self.raft_proto_handle.synchronize()
+
+        return preds
+
+    def predict_per_tree(
+            self,
```

Comment:
We should also optionally allow the user to provide an output array.

Reply:
Done